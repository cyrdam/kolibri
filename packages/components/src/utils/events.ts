// TODO: Should be synchronized with enums/events.ts
export type KoliBriEventType =
	| 'blur'
	| 'change'
	| 'changePage'
	| 'changePageSize'
	| 'click'
	| 'close'
	| 'focus'
	| 'input'
	| 'mousedown'
	| 'reset'
	| 'selection-change' // todo remove
	| 'selectionChange'
	| 'sort'
	| 'submit'
	| 'toggle';

export function stopPropagation(event: Event): void {
	event.stopImmediatePropagation();
	event.stopPropagation();
}

function createKoliBriEvent<T>(type: KoliBriEventType, detail: T): CustomEvent {
	const event = new CustomEvent(`kol-${type}`, {
		bubbles: true,
		cancelable: true,
		composed: true,
		detail: detail,
	});
	return event;
}

function dispatchKoliBriEvent<T>(target: EventTarget, type: KoliBriEventType, detail?: T): boolean {
	const dispatch = target.dispatchEvent(createKoliBriEvent(type, detail));
	return dispatch;
}

/**
 * @deprecated
 */
export function tryToDispatchKoliBriEvent<T>(type: KoliBriEventType, target?: EventTarget, detail?: T): void {
	target && dispatchKoliBriEvent(target, type, detail);
}

export function dispatchDomEvent(target: HTMLElement, type: KoliBriEventType) {
	target.dispatchEvent(new Event(type, { bubbles: true, composed: true }));
}

export function preventDefaultAndStopPropagation(event: Event) {
	event.preventDefault();
	stopPropagation(event);
}
