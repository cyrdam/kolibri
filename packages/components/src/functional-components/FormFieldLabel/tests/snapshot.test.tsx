import { h } from '@stencil/core';
import KolFormFieldLabelFc from '../FormFieldLabel';
import { renderFunctionalComponentToSpecPage } from '../../../utils/testing';

describe('KolFormFieldLabelFc', () => {
	it('should render with label', async () => {
		const label = 'Test Label';
		const id = 'test-id';
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldLabelFc label={label} id={id} />);

		expect(page.root).toMatchSnapshot();
	});

	it('should render with accessKey and shortKey', async () => {
		const label = 'Test Label';
		const accessKey = 'A';
		const shortKey = 'S';
		const id = 'test-id';
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldLabelFc label={label} accessKey={accessKey} shortKey={shortKey} id={id} />);

		expect(page.root).toMatchSnapshot();
	});

	it('should render with expert slot', async () => {
		const id = 'test-id';
		const page = await renderFunctionalComponentToSpecPage(() => (
			<KolFormFieldLabelFc id={id} hasExpertSlot>
				<slot name="expert">Expert Content</slot>
			</KolFormFieldLabelFc>
		));

		expect(page.root).toMatchSnapshot();
	});

	it('should set the correct id and class names', async () => {
		const label = 'Test Label';
		const id = 'test-id';
		const classNames = 'custom-class';
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldLabelFc label={label} id={id} class={classNames} />);

		expect(page.root?.id).toBe(`${id}-label`);
		expect(page.root?.className).toContain('input-label');
		expect(page.root?.className).toContain(classNames);
	});

	it('should hide label when hideLabel is true', async () => {
		const label = 'Test Label';
		const id = 'test-id';
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldLabelFc label={label} id={id} hideLabel />);

		expect(page.root?.hidden).toBe(true);
	});

	it('should render the correct component tag "legend" ', async () => {
		const label = 'Test Label';
		const id = 'test-id';
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldLabelFc label={label} id={id} component="legend" />);

		expect(page.root?.tagName).toBe('LEGEND');
		expect(page.root).toMatchSnapshot();
	});

	it('should render the correct component tag "label" ', async () => {
		const label = 'Test Label';
		const id = 'test-id';
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldLabelFc label={label} id={id} component="label" />);

		expect(page.root?.tagName).toBe('LABEL');
		expect(page.root).toMatchSnapshot();
	});
});
