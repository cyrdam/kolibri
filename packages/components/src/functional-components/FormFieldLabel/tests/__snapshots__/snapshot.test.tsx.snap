// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KolFormFieldLabelFc should render the correct component tag "label"  1`] = `
<label class="input-label" htmlfor="test-id" id="test-id-label">
  <span class="input-label-span">
    <span>
      Test Label
    </span>
  </span>
</label>
`;

exports[`KolFormFieldLabelFc should render the correct component tag "legend"  1`] = `
<legend class="input-label" htmlfor="test-id" id="test-id-label">
  <span class="input-label-span">
    <span>
      Test Label
    </span>
  </span>
</legend>
`;

exports[`KolFormFieldLabelFc should render with accessKey and shortKey 1`] = `
<label class="input-label" htmlfor="test-id" id="test-id-label">
  <span class="input-label-span">
    Test Label
    <span aria-hidden="true" class="access-key-hint">
      A
    </span>
  </span>
</label>
`;

exports[`KolFormFieldLabelFc should render with expert slot 1`] = `
<label class="input-label" htmlfor="test-id" id="test-id-label">
  <span class="input-label-span"></span>
</label>
`;

exports[`KolFormFieldLabelFc should render with label 1`] = `
<label class="input-label" htmlfor="test-id" id="test-id-label">
  <span class="input-label-span">
    <span>
      Test Label
    </span>
  </span>
</label>
`;
